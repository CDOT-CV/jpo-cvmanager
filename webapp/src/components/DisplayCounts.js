import React, { useState, useEffect } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { DateTimePickerComponent } from '@syncfusion/ej2-react-calendars'
import { MessageTypes } from '../constants/messageTypes'
import BounceLoader from 'react-spinners/BounceLoader'
import Select from 'react-select'
import {
    selectRequestOut,
    selectMsgType,
    selectCountList,
    selectStartDate,
    selectEndDate,
    selectWarningMessage,
    selectMessageLoading,
    updateRowData,
    updateMessageType,
} from '../slices/rsuSlice'

import './css/SnmpwalkMenu.css'

const messageTypeOptions = MessageTypes.map((type) => {
    return { value: type, label: type }
})
const { DateTime } = require('luxon')
const DisplayCounts = (props) => {
    const dispatch = useDispatch()
    const msgType = useSelector(selectMsgType)
    const startDate = useSelector(selectStartDate)
    const endDate = useSelector(selectEndDate)
    const requestOut = useSelector(selectRequestOut)
    const [previousRequest, setPreviousRequest] = useState(null)
    const warning = useSelector(selectWarningMessage)
    const messageLoading = useSelector(selectMessageLoading)
    const countList = useSelector(selectCountList)
    const [currentSort, setCurrentSort] = useState(null)
    const [sortedCountList, setSortedCountList] = useState(countList)
    const dateChanged = (e, type) => {
        let tmp = new Date(e.value.toString().substring(0, 24))
        let mst = DateTime.fromISO(tmp.toISOString())
        mst.setZone('America/Denver')
        let data
        if (type === 'start') {
            data = { start: mst.toString() }
        } else {
            data = { end: mst.toString() }
        }
        updateRowDataLocal(data)
    }

    useEffect(() => {
        setSortedCountList(countList)
    }, [countList])

    const getWarningMessage = (warning) =>
        warning ? (
            <span className="warningMessage">
                <p>
                    Warning: time ranges greater than 24 hours may have longer
                    load times.
                </p>
            </span>
        ) : (
            <span></span>
        )
    const updateRowDataLocal = (data) => {
        if (requestOut) {
            previousRequest.abort()
            setPreviousRequest(null)
        }
        dispatch(updateRowData(data))
    }
    const compareBy = (key) => {
        // Support both descending and ascending sort
        // based on the current sort
        // Default is ascending
        if (key === currentSort) {
            setCurrentSort(key + 'desc')
            return function (a, b) {
                if (a[key] > b[key]) return -1
                if (a[key] < b[key]) return 1
                return 0
            }
        } else {
            setCurrentSort(key)
            return function (a, b) {
                if (a[key] < b[key]) return -1
                if (a[key] > b[key]) return 1
                return 0
            }
        }
    }
    const sortBy = (key) => {
        let arrayCopy = [...countList]
        arrayCopy.sort(compareBy(key))
        setSortedCountList(arrayCopy)
    }
    const getTable = (messageLoading, sortedCountList) =>
        messageLoading ? (
            <div>
                <div className="table">
                    <div className="header">
                        <div>RSU</div>
                        <div>Road</div>
                        <div>Count</div>
                    </div>
                </div>
                <span className="bounceLoader">
                    <BounceLoader
                        loading={true}
                        color={'#ffffff'}
                    ></BounceLoader>
                </span>
            </div>
        ) : (
            <div className="table">
                <div className="header">
                    <div onClick={() => sortBy('rsu')}>RSU</div>
                    <div onClick={() => sortBy('road')}>Road</div>
                    <div onClick={() => sortBy('count')}>Count</div>
                </div>
                <div className="body">{formatRows(sortedCountList)}</div>
            </div>
        )
    const formatRows = (rows) => rows.map((rowData) => <Row {...rowData} />)
    return (
        <div>
            <div id="container" className="sideBarOn">
                <h1 className="h1">{msgType} Counts</h1>
                <div className="DateRangeContainer">
                    <DateTimePickerComponent
                        placeholder="Select start date"
                        value={startDate}
                        max={endDate}
                        step={15}
                        onChange={(e) => {
                            dateChanged(e, 'start')
                        }}
                    />
                    <DateTimePickerComponent
                        placeholder="Select end date"
                        value={endDate}
                        min={startDate}
                        max={new Date()}
                        step={15}
                        onChange={(e) => {
                            dateChanged(e, 'end')
                        }}
                    />
                </div>
                <Select
                    options={messageTypeOptions}
                    defaultValue={messageTypeOptions.filter(
                        (o) => o.label === msgType
                    )}
                    placeholder="Select Message Type"
                    className="selectContainer"
                    onChange={(value) =>
                        dispatch(updateMessageType(value.value))
                    }
                />
                {getWarningMessage(warning)}
                {getTable(messageLoading, sortedCountList)}
            </div>
        </div>
    )
}
const Row = ({ rsu, road, count }) => (
    <div className="row">
        <div>{rsu}</div>
        <div>{road}</div>
        <div>{count}</div>
    </div>
)
export default DisplayCounts
