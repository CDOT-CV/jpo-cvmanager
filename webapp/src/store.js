import { configureStore } from "@reduxjs/toolkit";
import rsuReducer from "./slices/rsuSlice";
import userReducer from "./slices/userSlice";
import wzdxReducer from "./slices/wzdxSlice";
import configReducer from "./slices/configSlice";

const store = configureStore({
  reducer: {
    rsu: rsuReducer,
    user: userReducer,
    wzdx: wzdxReducer,
    config: configReducer,
  },
});

export default store;
