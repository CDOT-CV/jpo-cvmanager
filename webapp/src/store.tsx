import { Action, ThunkAction, configureStore } from '@reduxjs/toolkit'
import rsuReducer from './generalSlices/rsuSlice'
import userReducer from './generalSlices/userSlice'
import wzdxReducer from './generalSlices/wzdxSlice'
import configReducer from './generalSlices/configSlice'
import adminAddOrganizationReducer from './features/adminAddOrganization/adminAddOrganizationSlice'
import adminAddRsuReducer from './features/adminAddRsu/adminAddRsuSlice'
import adminAddUserReducer from './features/adminAdduser/adminAddUserSlice'
import adminEditOrganizationReducer from './features/adminEditOrganization/adminEditOrganizationSlice'
import adminEditRsuReducer from './features/adminEditRsu/adminEditRsuSlice'
import adminEditUserReducer from './features/adminEditUser/adminEditUserSlice'
import adminOrganizationTabReducer from './features/adminOrganizationTab/adminOrganizationTabSlice'
import adminOrganizationTabUserReducer from './features/adminOrganizationTabUser/adminOrganizationTabUserSlice'
import adminOrganizationTabRsuReducer from './features/adminOrganizationTabRsu/adminOrganizationTabRsuSlice'
import adminRsuTabReducer from './features/adminRsuTab/adminRsuTabSlice'
import adminUserTabReducer from './features/adminUserTab/adminUserTabSlice'
import menuReducer from './features/menu/menuSlice'
import intersectionMapReducer from './components/map/map-slice'
import intersectionMapLayerStyleReducer from './components/map/map-layer-style-slice'

export const setupStore = (preloadedState: any) => {
  return configureStore({
    reducer: {
      rsu: rsuReducer,
      user: userReducer,
      wzdx: wzdxReducer,
      config: configReducer,
      adminAddOrganization: adminAddOrganizationReducer,
      adminAddRsu: adminAddRsuReducer,
      adminAddUser: adminAddUserReducer,
      adminEditOrganization: adminEditOrganizationReducer,
      adminEditRsu: adminEditRsuReducer,
      adminEditUser: adminEditUserReducer,
      adminOrganizationTab: adminOrganizationTabReducer,
      adminOrganizationTabUser: adminOrganizationTabUserReducer,
      adminOrganizationTabRsu: adminOrganizationTabRsuReducer,
      adminRsuTab: adminRsuTabReducer,
      adminUserTab: adminUserTabReducer,
      menu: menuReducer,
      intersectionMap: intersectionMapReducer,
      intersectionMapLayerStyle: intersectionMapLayerStyleReducer,
    },
    preloadedState,
  })
}

type AppStore = ReturnType<typeof setupStore>
export type AppState = ReturnType<AppStore['getState']>
type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, unknown, Action>

export type RootState = ReturnType<ReturnType<typeof setupStore>['getState']>
