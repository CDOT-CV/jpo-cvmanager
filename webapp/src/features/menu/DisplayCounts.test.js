import React from "react";
import { render } from "@testing-library/react";
import DisplayCounts from "./DisplayCounts";
import { Provider } from "react-redux";
import { setupStore } from "../../store";

it("should take a snapshot", () => {
  const { asFragment } = render(
    <Provider store={setupStore({})}>
      <DisplayCounts />
    </Provider>
  );

  expect(
    asFragment(
      <Provider store={setupStore({})}>
        <DisplayCounts />
      </Provider>
    )
  ).toMatchSnapshot();
});
