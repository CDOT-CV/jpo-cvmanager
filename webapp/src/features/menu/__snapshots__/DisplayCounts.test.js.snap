// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="sideBarOn"
      id="container"
    >
      <h1
        class="h1"
      >
        BSM Counts
      </h1>
      <div
        class="DateRangeContainer"
      >
        <div
          style="margin-top: 10px;"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="mui-1"
              id="mui-1-label"
            >
              Select start date
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="true"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                id="mui-1"
                placeholder="mm/dd/yyyy hh:mm (a|p)m"
                type="tel"
                value=""
              />
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
              >
                <button
                  aria-label="Choose date"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="CalendarIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Select start date
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          style="margin-top: 10px;"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="mui-2"
              id="mui-2-label"
            >
              Select end date
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="true"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                id="mui-2"
                placeholder="mm/dd/yyyy hh:mm (a|p)m"
                type="tel"
                value=""
              />
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
              >
                <button
                  aria-label="Choose date"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="CalendarIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Select end date
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        class="selectContainer css-b62m3t-container"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-2-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-1s2u09g-control"
        >
          <div
            class=" css-319lph-ValueContainer"
          >
            <div
              class=" css-qc6sy-singleValue"
            >
              BSM
            </div>
            <div
              class=" css-6j8wv5-Input"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-2-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              class=" css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-tlfecz-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <span />
      <div
        class="table"
      >
        <div
          class="header"
        >
          <div>
            RSU
          </div>
          <div>
            Road
          </div>
          <div>
            Count
          </div>
        </div>
        <div
          class="body"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
